service/raft/group0_state_machine.cc

```diff
@@ -108,6 +108,9 @@ future<> group0_state_machine::apply(std::vector<raft::command_cref> command) {
         },
         [&] (topology_change& chng) -> future<> {
            return _ss.topology_transition(_sp, _cdc_gen_svc, cmd.creator_addr, std::move(chng.mutations));
+        },
+        [&] (write_mutations& muts) -> future<> {
+           return _ss.write_topology_mutations(_sp, cmd.creator_addr, std::move(muts.mutations));
```

Comment:
The mutations sent in `write_mutations` don't necessarily have to be "topology mutations". In fact the mutations that you're sending with this command in later commits - I wouldn't really call them topology mutations, they don't affect the `topology` struct or `system.topology` table. They write data to CDC generations table (`system.cdc_generations`), and only the last command updates `topology` to point to that data.

Furthermore, it's not safe in general to split topology mutations (those that may modify `topology` struct) and commit them to group 0 in multiple commands. Because after applying once of them, a node may crash and then restart. Then it will call `topology_state_load` and observe partial state.

The CDC mutations don't have this problem. If we crash after committing only a subset of them, it won't affect anything other than storing some garbage data in `system.cdc_generations` that isn't referenced by anything. Only the last command sent by the coordinator updates the `system.topology` table and the `topology` struct which stores a pointer (key) to `system.cdc_generations` which contains the entire generation data at this moment.

So in fact the mutations sent by `write_mutations` not only aren't necessarily topology mutations - they must not be topology mutations, unless they contain a full update of topology state after which it's safe to call `topology_state_load`. But then you'd use `topology_change` command instead of `write_mutations`. So `write_mutations` are for purposes different than updating topology.

Having said all that, I propose the following changes:
- write a comment at `struct write_mutations` (in group0_state_machine.hh) explaining that this command is used to write data to tables other than topology or schema tables and it doesn't update any in-memory data structures
- rename `write_topology_mutations` function to something more generic, maybe simply `write_mutations`. I would also put it outside `storage_service`, maybe in group0_state_machine.cc (it looks like it could be a free-standing function, not a member). Then `storage_service::topology_transition` won't be able to call it, so put the call also in group0_state_machine in the `topology_change` case above the call to `topology_transition`. You should then specify at `topology_transition` declaration site the precondition that the topology mutations were already written to disk; the function only transitions the in-memory state machine.

Reply:
Done.